### PR TITLE
meshoptimizer: init at unstable-2023-03-22

### LINKED ISSUE #223406: Package request: gltfpack
**Project description**

gltfpack is an application that can automatically optimise glTF files to reduce their complexity and file size. glTF is a open standard, created by Khronos, for describing 3D scenes.

**Metadata**

* homepage URL: https://github.com/zeux/meshoptimizer
* source URL: https://github.com/zeux/meshoptimizer
* license: mit
* platforms: linux, darwin, windows

### PR DESCRIPTION
**Pull request by lillycham**
###### Description of changes
Added the package meshoptimizer, and its gltfpack binary.
packages homepage: https://github.com/zeux/meshoptimizer/
Closes #223406 
###### Things done

Can't currently test it on linux, as my other machine is broken right now.

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/graphics/gltfpack/default.nix
Comment (nagy) [inline pkgs/applications/graphics/gltfpack/default.nix:45]: This needs a meta block as well.

File: pkgs/applications/graphics/gltfpack/default.nix
Comment (nagy) [inline pkgs/applications/graphics/gltfpack/default.nix:4]: You are not using this argument, so it can be removed. In the other package as well.

File: pkgs/applications/graphics/gltfpack/default.nix
Comment (nagy) [inline pkgs/applications/graphics/gltfpack/default.nix:23]: Please use `rev = "refs/tags/v${version}"` instead. This makes updating easier.  Same for the other package.

File: pkgs/applications/graphics/gltfpack/default.nix
Comment (nagy) [inline pkgs/applications/graphics/gltfpack/default.nix:17]: This should be `pname`. Same for the other package.

File: pkgs/applications/graphics/gltfpack/default.nix
Comment (nagy) [inline pkgs/applications/graphics/gltfpack/default.nix:26]: if `meshoptimizer` is needed at runtime, then we have to put it in `buildInputs`. `cmake` can stay here, since it is only needed at build time.

**Review by nagy — CHANGES_REQUESTED** [review]
You are overwriting a few phases. In general it is best not to overwrite them directly because the internal default one are running hooks like `preConfigure` and `postConfigure`. That makes it easy to extend from outside. Let me see If I can get this to build without phase declarations.

After all these changes your PR should consist of 3 Commits, one for the init of `meshoptimizer` one for the init of `gltfpack` and your maintainership commit.

What you could also do is to combine the two new packages and have the `gltfpack` binary be installed in the `bin` output of the `meshoptimizer` package and its developmentfiles can be installed in the `dev` output. Most of that should be handled automatically if you just declare òutputs = ["bin" "dev"];` , but that is up to you.

**Comment by nagy** [comment]
I can confirm, that it works on `x86_64-linux`.

File: pkgs/applications/graphics/gltfpack/default.nix
Comment (lillycat332) [reply pkgs/applications/graphics/gltfpack/default.nix:23]: I did, but it wouldn't compile w/the binary output enabled when i used a specific tag.

Comment (lillycat332) [reply pkgs/applications/graphics/gltfpack/default.nix:23]: Is it acceptable to downgrade the version to 0.17? I can't actually get the 0.18 to build reproducibly w/a binary output on my machine.

Comment (nagy) [reply pkgs/applications/graphics/gltfpack/default.nix:23]: I can confirm, that the v0.18 version does not compile. Unfortunately, the v0.18 version is very old and no new release is available. https://github.com/zeux/meshoptimizer/compare/v0.18...49d9222 . Maybe we should set the `version` attribute to `"unstable-2023-03-22"` then, because that is the day of the 49d9222 commit.

Comment (nagy) [reply pkgs/applications/graphics/gltfpack/default.nix:23]: > Is it acceptable to downgrade the version to 0.17? I can't actually get the 0.18 to build reproducibly w/a binary output on my machine.

We don't have to downgrade. We can use this commit with the unstable version attribute.

**Comment by nagy** [comment]
>  Should solve #223406.

Also, if you use one of these described keywords over here https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword then the accompanying issue will be closed automatically, once this PR is merged.

**Comment by lillycat332** [comment]
Sorry, I can't figure out how to squash my commits together here (I think because I pulled from upstream in between it messed it up?)

**Comment by nagy** [comment]
If you have this checkbox "Allow edits from maintainers" in the sidebox ticked, I might be able to fix the commits on your branch.

**Comment by lillycat332** [comment]
Yes, I fixed that on my local copy :) just didn't want to push until I sorted out the commits

File: pkgs/development/libraries/meshoptimizer/default.nix
Comment (nagy) [inline pkgs/development/libraries/meshoptimizer/default.nix:22]: I have noticed, that the "out" output is empty. I would probably go with this approach, that is documented here: https://github.com/NixOS/nixpkgs/blob/22.11/doc/stdenv/multiple-output.chapter.md#writing-a-split-derivation-sec-multiple-outputs-

So that means changing it to:  `outputs = [ "bin" "dev" "out" ];`

**Comment by nagy** [comment]
After these last two changes, it should be fine.

**Review by nagy — APPROVED** [review]

**Review by NickCao — APPROVED** [review]

**Pull request merged**
